### PR TITLE
Package vast-bridge as PyPI package

### PR DESCRIPTION
**Pull request by 0ortmann**
###  :notebook_with_decorative_cover: Description

- Rename `vast-bridge` to `pyvast-threatbus` as per discussion in Slack
- Package as PyPI package
- Add a CHANGELOG to the app to track it as an individual project

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

File-by-file or commit by commit, up to taste.
Test the module as follows:

- navigate to `apps/vast`
- run `make dist`
- setup a virtual environment, somewhere on your computer, via `virtualenv venv`
- activate it `source venv/bin/activate`
- install the freshly built module: `pip install /path/to/threatbus/apps/vast/dist/pyvast-threatbus-2020.11.26.tar.gz`
- run it via `venv/bin/pyvast-threatbus -c apps/vast/config.yaml`

### CODE REVIEW

File: apps/vast/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Changelog
+
+This changelog documents all notable user-facing changes of `pyvast-threatbus`.
+
+Every entry has a category for which we use the following visual abbreviations:
+
+- 🎁 feature
+- 🧬 experimental feature
+- ⚠️ change
+- 🐞 bugfix
```
Comment (mavam) [inline apps/vast/CHANGELOG.md:10]: We could use our latest VAST legend that also includes a breaking change.

File: apps/vast/README.md [hunk 2]
```diff
@@ -1,58 +1,70 @@
...
 
-- Install the `requirements.txt` for the VAST wrapper script
+Install `pyvast-threatbus` via `pip`. Optionally, use a virtual environment.
 
-```sh
-virtualenv --system-site-packages venv
+```
+virtualenv venv           # optional
+source venv/bin/activate  # optional
+python3 -m pip install pyvast-threatbus
```
Comment (mavam) [inline apps/vast/README.md:20]:
```suggestion
python -m pip install pyvast-threatbus
```

**Review by mavam — APPROVED** [review]
LGTM.

Just two small mechanical change requests.

I don't know how to test the pypi package, but once merged, I'll try to install. 🙂

**Comment by 0ortmann** [comment]
@mavam I updated the PR instructions for how to test a Python module. Please let me know if that works on your machine.

**Comment by mavam** [comment]
Thanks, the instructions are great. Unfortunately I'm running into an error when trying to install.

```
python3 -m venv venv
source venv/bin/activate
pip install dist/pyvast-threatbus-2020.11.26.tar.gz
Processing ./dist/pyvast-threatbus-2020.11.26.tar.gz
Processing /Users/mavam/Library/Caches/pip/wheels/95/a4/59/10cd5378d52f92cdb45025f040e4686e10ae5217961c25fd66/black-20.8b1-py3-none-any.whl
Collecting coloredlogs>=14.0
  Downloading coloredlogs-14.0-py2.py3-none-any.whl (43 kB)
     |████████████████████████████████| 43 kB 5.1 MB/s
Collecting confuse
  Downloading confuse-1.3.0-py2.py3-none-any.whl (64 kB)
     |████████████████████████████████| 64 kB 3.7 MB/s
Collecting pyzmq>=19
  Downloading pyzmq-19.0.2-cp38-cp38-macosx_10_9_x86_64.whl (806 kB)
     |████████████████████████████████| 806 kB 9.0 MB/s
Collecting pyvast>=2020.10.29
  Using cached pyvast-2020.10.29-py3-none-any.whl (5.1 kB)
ERROR: Could not find a version that satisfies the requirement threatbus>=2020.11.26 (from pyvast-threatbus==2020.11.26) (from versions: 0.3.0, 0.3.1, 0.3.2, 2020.1.31, 2020.2.27, 2020.4.29, 2020.5.28, 2020.6.25, 2020.7.28, 2020.9.30, 2020.10.29)
ERROR: No matching distribution found for threatbus>=2020.11.26 (from pyvast-threatbus==2020.11.26)
WARNING: You are using pip version 20.2.1; however, version 20.2.4 is available.
You should consider upgrading via the '/Users/mavam/code/tenzir/threatbus/apps/vast/venv/bin/python3 -m pip install --upgrade pip' command.
```

Thereafter I tried `pip install threatbus`, which ran through, but the above error still persists.

**Comment by 0ortmann** [comment]
Yeah, you need to install Threat Bus from this repo.The latest version of the `threatbus` package on PyPI is 2020.10.29. You can do that the same way you install this app or you use development mode, e.g., `python setup.py develop` in the top-level folder.

**Comment by mavam** [comment]
Here's the latest error message when calling `venv/bin/pyvast-threatbus -h`:

```
Traceback (most recent call last):
  File "venv/bin/pyvast-threatbus", line 33, in <module>
    sys.exit(load_entry_point('pyvast-threatbus==2020.11.26', 'console_scripts', 'pyvast-threatbus')())
  File "venv/bin/pyvast-threatbus", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/local/Cellar/python@3.8/3.8.6_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/usr/local/Cellar/python@3.8/3.8.6_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'pyvast_threatbus'
```

**Comment by 0ortmann** [comment]
@mavam please re-test this.

**Comment by mavam** [comment]
1. `cd apps/vast`
2. `make dist`
3. `python3 -m venv venv`
4. `source venv/bin/activate`
5. `pip install dist/pyvast-threatbus-2020.11.26.tar.gz`

    ```
    ERROR: Could not find a version that satisfies the requirement threatbus>=2020.11.26 (from pyvast-threatbus==2020.11.26) (from versions: 0.3.0, 0.3.1, 0.3.2, 2020.1.31, 2020.2.27, 2020.4.29, 2020.5.28, 2020.6.25, 2020.7.28, 2020.9.30, 2020.10.29)
    ERROR: No matching distribution found for threatbus>=2020.11.26 (from pyvast-threatbus==2020.11.26)
    ```
6. `pip install threatbus`

    Same error.

7. `pushd ../../` (top-level repo dir)
8. `make dist`

    ```
    make[1]: [clean] Error 1 (ignored)
    rm -f -r build
    python plugins/apps/threatbus_misp/setup.py sdist bdist_wheel
    usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
       or: setup.py --help [cmd1 cmd2 ...]
       or: setup.py --help-commands
       or: setup.py cmd --help

    error: invalid command 'bdist_wheel'
    make: *** [dist] Error 1
    ```

Summary: it looks like a certain `threatbus` version is needed. When trying to get a compatible version from the top-level repo via `make dist`, I am running into the above error.

**Comment by 0ortmann** [comment]
@mavam Before attempting to build you must install build-dependencies. I.e., for the `bdist_wheel` target, you need the `wheel` package.

**Comment by mavam** [comment]
> @mavam Before attempting to build you must install build-dependencies. I.e., for the bdist_wheel target, you need the wheel package.

Shouldn't dependencies be installed automatically? (At least the review instructions didn't include a special dependency setup for the `venv`.)

**Comment by 0ortmann** [comment]
It's not in the review instructions because I assumed that the person reviewing this would have built at least some python package in the past already or at least know that `wheel` needs to be present for building `wheel` packges. It cannot be installed automagically, that is an ongoing discussion, see [here](https://www.python.org/dev/peps/pep-0518/#rationale) why. We do list it in our `setup_requires` section.

So in order to move forward: have you installed the build dependency and tested this package?

**Comment by mavam** [comment]
> So in order to move forward: have you installed the build dependency and tested this package?

This is what I'm trying; I'm not getting very far. My list of steps should tell you exactly where I'm at. Sorry, I'm a Python packaging noob. If you let me know what extra commands or packages I need, I'll execute them. I did `pip install wheel` and the warning is no longer here. No the output is only littered with no-such-file-or-directory warnings (errors?)

I'm still trying to get `threatbus` installed locally, via `make dist`. This is the output:

```
python setup.py sdist bdist_wheel
running sdist
running egg_info
creating threatbus.egg-info
writing threatbus.egg-info/PKG-INFO
writing dependency_links to threatbus.egg-info/dependency_links.txt
writing entry points to threatbus.egg-info/entry_points.txt
writing requirements to threatbus.egg-info/requires.txt
writing top-level names to threatbus.egg-info/top_level.txt
writing manifest file 'threatbus.egg-info/SOURCES.txt'
reading manifest file 'threatbus.egg-info/SOURCES.txt'
writing manifest file 'threatbus.egg-info/SOURCES.txt'
running check
creating threatbus-2020.11.26
...
make clean
find . -type d -name "*egg-info" -exec rm -f -r {} \;
find: ./apps/vast/pyvast_threatbus.egg-info: No such file or directory
find: ./apps/vast/build/lib/pyvast_threatbus.egg-info: No such file or directory
make[1]: [clean] Error 1 (ignored)
find . -type d -name "__pycache__" -exec rm -f -r {} \;
find: ./apps/vast/venv/lib/python3.8/site-packages/wheel/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/wheel/cli/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/wheel/vendored/packaging/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/wheel/vendored/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/setuptools/_vendor/packaging/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/setuptools/_vendor/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/setuptools/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/setuptools/command/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/setuptools/extern/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/pkg_resources/_vendor/packaging/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/pkg_resources/_vendor/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/pkg_resources/__pycache__: No such file or directory
find: ./apps/vast/venv/lib/python3.8/site-packages/pkg_resources/extern/__pycache__: No such file or directory
make[1]: [clean] Error 1 (ignored)
rm -f -r build
```

In general, can we fail hard if requirements are not met? Then it would be obvious what's going on. 

Anyhow, I got some files to work with:

```
ls dist
threatbus-2020.11.26-py3-none-any.whl  threatbus-misp-2020.10.29.tar.gz      threatbus_cif3-2020.10.29-py3-none-any.whl      threatbus_zeek-2020.10.29-py3-none-any.whl
threatbus-2020.11.26.tar.gz            threatbus-rabbitmq-2020.10.29.tar.gz  threatbus_inmem-2020.10.29-py3-none-any.whl     threatbus_zmq_app-2020.10.29-py3-none-any.whl
threatbus-cif3-2020.10.29.tar.gz       threatbus-zeek-2020.10.29.tar.gz      threatbus_misp-2020.10.29-py3-none-any.whl
threatbus-inmem-2020.10.29.tar.gz      threatbus-zmq-app-2020.10.29.tar.gz   threatbus_rabbitmq-2020.10.29-py3-none-any.whl
```

I then did a `pip install dist/threatbus-2020.11.26.tar.gz`, which seemed to work. Thereafter, `pip install dist/pyvast-threatbus-2020.11.26.tar.gz` also worked.

Then I was wondering how the config should look like for testing. I don't have a MISP currently up and running. I have only the `zmq-app` activated under `apps` in the YAML file. When I start `threatbus -c config.yaml`, I get this output only:

```
threatbus -c config.yaml
2020-11-16 14:30:44 INFO     [threatbus] Starting plugins...

```

Is there anything missing? It seems that the bridge cannot find Threat Bus:

```
venv/bin/pyvast-threatbus -c config.yaml
2020-11-16 14:34:09 INFO     PyVAST: VAST client configured to use endpoint localhost:42000
2020-11-16 14:34:09 DEBUG    Calling Threat Bus management endpoint localhost:13370
2020-11-16 14:34:09 INFO     Subscribing to topic 'threatbus/intel'...
2020-11-16 14:34:14 ERROR    Subscription failed
```

**Comment by 0ortmann** [comment]
That test completes this story. Both packages could be built, the build artifacts could be installed. Hence, they both form valid packages and I'll merge this PR :tada: 

Explanation of the log messages:

- `make dist` has successfully produced the Python package artifacts
- The `no such file or directory` logs you see come from the `clean` target (Makefile), i.e., when `find` does not find all the patterns we want to clean.
- we do fail hard if requirements are not met, except for cleaning
- see `config.yaml.example` for an example config

From the two application logs:

1. There are no threatbus plugins installed, otherwise they would pop up in the log. Make sure you install them, e.g., via `pip install dist/some-plugin.tgz`. Your config won't do anything if you don't install the plugins.
2. `pyvast-threatbus` ideed cannot find Threat Bus. That is because the `zmq-app` plugin is not installed (see 1.)

**Comment by mavam** [comment]
Thanks for the explanations. This makes sense now. I have a suggestion regarding UX: how about printing a warning if we have a config key but no corresponding plugin installed?

**Comment by 0ortmann** [comment]
Sure, we can do that. Can you make a story for it, please?

**Pull request merged**